File: backend/src/main/java/com/example/backend/model/NotificationTemplate.java
package com.example.backend.model;

import com.example.backend.common.NotificationType;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

@Entity
@Table(name = "notification_templates")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class NotificationTemplate {
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    @Column(name = "template_id")
    private UUID id;

    @Column(name = "name", nullable = false, length = 100)
    private String name;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false, length = 20)
    private NotificationType type;

    @Column(name = "subject")
    private String subject;

    // FreeMarker template path file
    @Column(name = "template_file", nullable = false)
    private String templateFile;

    @Column(name = "default_language", length = 10)
    private String defaultLanguage;

    @Builder.Default
    @Column(name = "priority")
    private Integer priority = 0;

    @Builder.Default
    @Column(name = "is_active")
    private Boolean isActive = true;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    // 1: N relationship with NotificationLog
    @OneToMany(mappedBy = "template", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<NotificationLog> logs = new ArrayList<>();
}

File: frontend/src/types/notificationTemplate.ts
import type { BaseEntity, BaseFilterRequest } from "./common";
import type { LanguageCode, NotificationType } from "./enum";

export interface NotificationTemplateResponse extends BaseEntity {
  name: string;
  type: NotificationType;
  subject: string;
  templateFile: string;
  defaultLanguage: LanguageCode;
  priority: number;
}

export interface NotificationTemplateRequest {
  name: string;
  type: NotificationType;
  subject?: string;
  templateFile: string;
  defaultLanguage?: LanguageCode;
  priority?: number;
}

export interface NotificationTemplateFilterRequest extends BaseFilterRequest {
  name?: string;
  type?: NotificationType;
  subject?: string;
  defaultLanguage?: LanguageCode;
  priorityMin?: number;
  priorityMax?: number;
}
